catroid/src/main/java/org/catrobat/catroid/drone/jumpingsumo/JumpingSumoDiscoverer.java

```diff
@@ -0,0 +1,188 @@
+		 */
+		void onDronesListUpdated(List<ARDiscoveryDeviceService> dronesList);
+	}
+
+	private final List<Listener> listeners;
+	private final Context context;
+	private ARDiscoveryService ardiscoveryService;
+	private ServiceConnection ardiscoveryServiceConnection;
+	private final ARDiscoveryServicesDevicesListUpdatedReceiver ardiscoveryServicesDevicesListUpdatedReceiver;
+	private final List<ARDiscoveryDeviceService> matchingDrones;
+
+	private boolean startDiscoveryAfterConnection = true;
+
+	public JumpingSumoDiscoverer(Context contextDrone) {
+		context = contextDrone;
+		listeners = new ArrayList<>();
+		matchingDrones = new ArrayList<>();
+		ardiscoveryServicesDevicesListUpdatedReceiver = new ARDiscoveryServicesDevicesListUpdatedReceiver(discoveryListener);
+	}
+	/*
+	* Add a listener
+	* All callbacks of the interface Listener will be called within this function
+	* Should be called in the main thread
+	* @param listener an object that implements the {@link Listener} interface
+	*/
+	public void addListener(Listener listener) {
+		listeners.add(listener);
+		notifyServiceDiscovered(matchingDrones);
+	}
+	/**
+	 * remove a listener from the listener list
+	 * @param listener an object that implements the {@link Listener} interface
+	 */
+
+	public void removeListener(Listener listener) {
+		listeners.remove(listener);
+	}
+
+	/**
+	 * Setup the drone discoverer
+	 * Should be called before starting discovering
+	 */
+	public void setup() {
+		LocalBroadcastManager localBroadcastMgr = LocalBroadcastManager.getInstance(context);
+		localBroadcastMgr.registerReceiver(ardiscoveryServicesDevicesListUpdatedReceiver,
+				new IntentFilter(ARDiscoveryService.kARDiscoveryServiceNotificationServicesDevicesListUpdated));
+
+		if (ardiscoveryServiceConnection == null) {
+			ardiscoveryServiceConnection = new ServiceConnection() {
+				@Override
+				public void onServiceConnected(ComponentName name, IBinder service) {
+					ardiscoveryService = ((ARDiscoveryService.LocalBinder) service).getService();
+					if (startDiscoveryAfterConnection) {
+						startDiscovering();
+						startDiscoveryAfterConnection = false;
+					}
+				}
+
+				@Override
+				public void onServiceDisconnected(ComponentName name) {
+					ardiscoveryService = null;
+				}
+			};
+		}
+
+		if (ardiscoveryService == null) {
+			Intent i = new Intent(context, ARDiscoveryService.class);
+			context.bindService(i, ardiscoveryServiceConnection, Context.BIND_AUTO_CREATE);
+		}
+	}
+
+	/**
+	 * Cleanup the object
+	 * Should be called when the object is not used anymore
+	 */
+	public void cleanup() {
+		stopDiscovering();
+		if (ardiscoveryService != null) {
+			new Thread(new Runnable() {
+				@Override
+				public void run() {
+					ardiscoveryService.stop();
+					context.unbindService(ardiscoveryServiceConnection);
+					ardiscoveryService = null;
+				}
+			}).start();
+		}
+
+		LocalBroadcastManager localBroadcastMgr = LocalBroadcastManager.getInstance(context);
```

Comment:
what is a "LocalBroadcastManager"?

Reply:
Helper to register for and send broadcasts of Intents to local objects within your process. 

We have one for register and also unregister an receiver. In setup we register an observer (ardiscoveryServicesDevicesListUpdatedReceiver) to receive Intents with actions named (ARDiscoveryService.kARDiscoveryServiceNotificationServicesDevicesListUpdated). To summerize what registerReceiver does: Register a receive for any local broadcasts that match the given IntentFilter.

Reply:
alright thanks, will be merged tomorrow ;)